### PR TITLE
Add class that supports dynamic reconfiguration of zookeeper

### PR DESCRIPTION
**Pull request by hmusum**
New class `ReconfigurableVespaZooKeeperServer` supports reconfiguration. Not used yet.

### CODE REVIEW

File: zookeeper-server/zookeeper-server-3.5.6/src/main/java/com/yahoo/vespa/zookeeper/Reconfigurer.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+import java.util.stream.Collectors;
+
+/**
+ * Starts zookeeper server and supports reconfiguring zookeeper cluster. Created as a component
+ * without any config injected, to make sure that it is not recreated when config changes.
+ *
+ * @author hmusum
+ */
+public class Reconfigurer extends AbstractComponent {
+    private static final java.util.logging.Logger log = java.util.logging.Logger.getLogger(Reconfigurer.class.getName());
```
Comment (mpolden) [inline zookeeper-server/zookeeper-server-3.5.6/src/main/java/com/yahoo/vespa/zookeeper/Reconfigurer.java:25]: Nit: Import `Logger`.

File: zookeeper-server/zookeeper-server-3.5.6/src/main/java/com/yahoo/vespa/zookeeper/Reconfigurer.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+    boolean shouldReconfigure(ZookeeperServerConfig newConfig) {
+        ZookeeperServerConfig existingConfig = zooKeeperRunner.zookeeperServerConfig();
+        if (!newConfig.dynamicReconfiguration() || existingConfig == null) return false;
+
+        return !newConfig.equals(existingConfig);
+    }
+
+    private ZooKeeperRunner startServer(ZookeeperServerConfig zookeeperServerConfig) {
+        //System.out.println("Starting server with config " + zookeeperServerConfig);
```
Comment (mpolden) [inline zookeeper-server/zookeeper-server-3.5.6/src/main/java/com/yahoo/vespa/zookeeper/Reconfigurer.java:51]: Remove.

File: zookeeper-server/zookeeper-server-3.5.6/src/main/java/com/yahoo/vespa/zookeeper/Reconfigurer.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+    private ZooKeeperRunner zooKeeperRunner;
+
+    @Inject
+    public Reconfigurer() {
+        log.log(Level.FINE, "Created ZooKeeperReconfigurer");
+    }
+
+    // For testing only
+    void startOrReconfigure(ZookeeperServerConfig newConfig) {
```
Comment (mpolden) [inline zookeeper-server/zookeeper-server-3.5.6/src/main/java/com/yahoo/vespa/zookeeper/Reconfigurer.java:35]: Stale comment? This is called from `ReconfigurableVespaZooKeeperServer`.

File: zookeeper-server/zookeeper-server-3.5.6/src/main/java/com/yahoo/vespa/zookeeper/Reconfigurer.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        }
+    }
+
+    List<String> currentServers() {
+        if (zooKeeperRunner == null) return List.of();
+
+        return servers(zooKeeperRunner.zookeeperServerConfig());
+    }
+
+    // Returns items in set a that are not in set b
```
Comment (mpolden) [inline zookeeper-server/zookeeper-server-3.5.6/src/main/java/com/yahoo/vespa/zookeeper/Reconfigurer.java:86]: Nit: Make it `/** javadoc */`.

File: zookeeper-server/zookeeper-server-3.5.6/src/test/java/com/yahoo/vespa/zookeeper/ReconfigurerTest.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+import java.io.IOException;
+import java.util.stream.IntStream;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Tests dynamic reconfiguration of zookeeper cluster.
+ */
+public class ReconfigurerTest {
```
Comment (mpolden) [inline zookeeper-server/zookeeper-server-3.5.6/src/test/java/com/yahoo/vespa/zookeeper/ReconfigurerTest.java:20]: Missing `@author`.

File: zookeeper-server/zookeeper-server-3.5.6/src/main/java/com/yahoo/vespa/zookeeper/ReconfigurableVespaZooKeeperServer.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+package com.yahoo.vespa.zookeeper;
+
+import com.google.inject.Inject;
+import com.yahoo.cloud.config.ZookeeperServerConfig;
+import com.yahoo.component.AbstractComponent;
+
+/**
+ * Starts or reconfigures zookeeper cluster
+ *
+ * @author Harald Musum
```
Comment (mpolden) [inline zookeeper-server/zookeeper-server-3.5.6/src/main/java/com/yahoo/vespa/zookeeper/ReconfigurableVespaZooKeeperServer.java:11]: Nit: `s/Harald Musum/hmusum/`?

File: zookeeper-server/zookeeper-server-3.5.6/src/main/java/com/yahoo/vespa/zookeeper/Reconfigurer.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+        List<String> joiningServers = servers(newConfig);
+        List<String> leavingServers = setDifference(originalServers, joiningServers);
+        List<String> addedServers = setDifference(joiningServers, leavingServers);
+
+        log.log(Level.INFO, "Will reconfigure zookeeper cluster. Joining servers: " + joiningServers +
+                            ", leaving servers: " + leavingServers +
+                            ", new members" + addedServers);
+
+        int sessionTimeoutInSeconds = 30;
```
Comment (mpolden) [inline zookeeper-server/zookeeper-server-3.5.6/src/main/java/com/yahoo/vespa/zookeeper/Reconfigurer.java:69]: Consider extracting as constant.

File: zookeeper-server/zookeeper-server-3.5.6/src/test/java/com/yahoo/vespa/zookeeper/ReconfigurerTest.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    @Test
+    public void testStartupAndReconfigure() {
+        Reconfigurer reconfigurer = new Reconfigurer();
+        reconfigurer.startOrReconfigure(createConfig(1));
+
+        // Created config has dynamicReconfig set to false
+        assertFalse(reconfigurer.shouldReconfigure(createConfig(2)));
+
+        // Created config has dynamicReconfig set to true
+        assertTrue(reconfigurer.shouldReconfigure(createConfigAllowReconfiguring(2)));
```
Comment (mpolden) [inline zookeeper-server/zookeeper-server-3.5.6/src/test/java/com/yahoo/vespa/zookeeper/ReconfigurerTest.java:43]: Consider testing that an equal config does not cause reconfiguration.

**Review by mpolden — COMMENTED** [review]
Looks good! I had a few minor comments.

File: zookeeper-server/zookeeper-server-3.5.6/src/main/java/com/yahoo/vespa/zookeeper/Reconfigurer.java (hunk 1 above)
Comment (hmusum) [reply zookeeper-server/zookeeper-server-3.5.6/src/main/java/com/yahoo/vespa/zookeeper/Reconfigurer.java]: Fixed

File: zookeeper-server/zookeeper-server-3.5.6/src/main/java/com/yahoo/vespa/zookeeper/Reconfigurer.java (hunk 3 above)
Comment (hmusum) [reply zookeeper-server/zookeeper-server-3.5.6/src/main/java/com/yahoo/vespa/zookeeper/Reconfigurer.java:36]: Fixed

File: zookeeper-server/zookeeper-server-3.5.6/src/main/java/com/yahoo/vespa/zookeeper/Reconfigurer.java (hunk 2 above)
Comment (hmusum) [reply zookeeper-server/zookeeper-server-3.5.6/src/main/java/com/yahoo/vespa/zookeeper/Reconfigurer.java]: Fixed

File: zookeeper-server/zookeeper-server-3.5.6/src/main/java/com/yahoo/vespa/zookeeper/Reconfigurer.java (hunk 4 above)
Comment (hmusum) [reply zookeeper-server/zookeeper-server-3.5.6/src/main/java/com/yahoo/vespa/zookeeper/Reconfigurer.java]: Fixed

File: zookeeper-server/zookeeper-server-3.5.6/src/main/java/com/yahoo/vespa/zookeeper/ReconfigurableVespaZooKeeperServer.java (hunk 6 above)
Comment (hmusum) [reply zookeeper-server/zookeeper-server-3.5.6/src/main/java/com/yahoo/vespa/zookeeper/ReconfigurableVespaZooKeeperServer.java]: Fixed

File: zookeeper-server/zookeeper-server-3.5.6/src/test/java/com/yahoo/vespa/zookeeper/ReconfigurerTest.java (hunk 5 above)
Comment (hmusum) [reply zookeeper-server/zookeeper-server-3.5.6/src/test/java/com/yahoo/vespa/zookeeper/ReconfigurerTest.java:22]: Fixed

File: zookeeper-server/zookeeper-server-3.5.6/src/main/java/com/yahoo/vespa/zookeeper/Reconfigurer.java (hunk 7 above)
Comment (hmusum) [reply zookeeper-server/zookeeper-server-3.5.6/src/main/java/com/yahoo/vespa/zookeeper/Reconfigurer.java]: Done

File: zookeeper-server/zookeeper-server-3.5.6/src/test/java/com/yahoo/vespa/zookeeper/ReconfigurerTest.java (hunk 8 above)
Comment (hmusum) [reply zookeeper-server/zookeeper-server-3.5.6/src/test/java/com/yahoo/vespa/zookeeper/ReconfigurerTest.java:45]: Good point, added

**Review by mpolden — APPROVED** [review]

**Pull request merged**